# Support Vector Machines

This model is an SVM (Support Vector Machine) classifier trained to distinguish facial emotions. SVM is a supervised learning technique that is used to perform classification and regression problems.

In [1]:
from sklearn.svm import SVC
from sklearn.model_selection import GridSearchCV
from sklearn.metrics import accuracy_score
import numpy as np
import pandas as pd

In [2]:
data = pd.read_csv('fer2013.csv')


In [3]:
train_data = data[data.Usage == 'Training']
test_data = data[data.Usage == 'PrivateTest']
val_data = data[data.Usage == 'PublicTest']


In [4]:
def preprocess(data):
    X = []
    y = []
    for i in range(len(data)):
        img = data.iloc[i]['pixels'].split(' ')
        img = np.array(img, dtype='float32')
        img = img / 255.0
        X.append(img)
        y.append(data.iloc[i]['emotion'])
    X = np.array(X)
    y = np.array(y)
    return X, y


In [5]:
X_train, y_train = preprocess(train_data)
X_val, y_val = preprocess(val_data)
X_test, y_test = preprocess(test_data)


In [6]:
X_train = X_train.reshape(X_train.shape[0], -1)
X_val = X_val.reshape(X_val.shape[0], -1)
X_test = X_test.reshape(X_test.shape[0], -1)


### SVM

The model is initiated using the SVC class from the sklearn.svm package to initialize the SVM classifier.

Parameters are selected using the gridsearch on lesser sample size dataset.

Key Parameters for SVM are Kernel, C.

In [10]:
from sklearn.svm import SVC

svm = SVC(C=1, kernel='rbf', random_state=42, verbose=True)
svm.fit(X_train, y_train)


[LibSVM]

SVC(C=1, random_state=42, verbose=True)

In [11]:
y_val_pred = svm.predict(X_val)
val_acc = accuracy_score(y_val, y_val_pred)
print("Validation accuracy:", val_acc)

Validation accuracy: 0.44664251880746725


In [12]:
y_test_pred = svm.predict(X_test)
test_acc = accuracy_score(y_test, y_test_pred)
print("Test accuracy:", test_acc)

Test accuracy: 0.4471997770966843


In [14]:
from sklearn.metrics import accuracy_score, classification_report
print("Accuracy:", accuracy_score(y_test, y_test_pred))
print(classification_report(y_test, y_test_pred))

Accuracy: 0.4471997770966843
              precision    recall  f1-score   support

           0       0.43      0.22      0.29       491
           1       1.00      0.05      0.10        55
           2       0.39      0.23      0.29       528
           3       0.48      0.72      0.58       879
           4       0.35      0.41      0.38       594
           5       0.65      0.53      0.58       416
           6       0.41      0.44      0.42       626

    accuracy                           0.45      3589
   macro avg       0.53      0.37      0.38      3589
weighted avg       0.45      0.45      0.43      3589



On the test data, the SVM classifier had an accuracy of 44.7%. Each emotion class had a different precision, recall, and F1-score. Disgust (1) and Surprise (5) exhibited worse precision and recall, indicating difficulty in accurately predicting these emotions. Emotion classes such as Happy (3) exhibited higher precision and recall, indicating that those emotions performed better. With a weighted average F1-score of 0.43, the model performed relatively poorly in reliably classifying emotions. To improve accuracy and overall performance in recognizing facial emotions, adjusting the SVM parameters or researching alternative models might be used.


In [15]:
import joblib

joblib.dump(svm, "svm_model.sav")


['svm_model.sav']